core/types/calls.cc

```diff
@@ -2395,20 +2396,52 @@ class Array_flatten : public IntrinsicMethod {
             // This only shows up because t->elementType(gs) for tuples returns an OrType of all its elements.
             // So to properly handle nested tuples, we have to descend into the OrType's.
             [&](const OrType &o) {
-                result = Types::any(gs, recursivelyFlattenArrays(gs, o.left, newDepth),
-                                    recursivelyFlattenArrays(gs, o.right, newDepth));
+                result = Types::any(gs, recursivelyFlattenArrays(gs, args, o.left, newDepth),
+                                    recursivelyFlattenArrays(gs, args, o.right, newDepth));
+            },
+
+            [&](const ClassType &c) {
+                // If the element type supports the #to_ary method, then Ruby will implicitly call it when
+                // flattening. So here we need to get the return type of calling #to_ary on an element.
+                DispatchArgs innerArgs{Names::toAry(), args.locs, 0,
+                                       args.args,      type,      {type, args.fullType.origins},
+                                       type,           nullptr,   args.originForUninitialized};
+
+                auto dispatched = c.dispatchCall(gs, innerArgs);
+                if (dispatched.main.errors.empty()) {
+                    result = recursivelyFlattenArrays(gs, args, move(dispatched.returnType), newDepth);
+                    return;
+                }
+
+                result = type;
             },
 
             [&](const AppliedType &a) {
-                if (a.klass != Symbols::Array()) {
-                    result = type;
+                // If we're dealing with a T::Array, then recursively flatten using the inner type.
+                if (a.klass == Symbols::Array()) {
+                    ENFORCE(a.targs.size() == 1);
+                    result = recursivelyFlattenArrays(gs, args, a.targs.front(), newDepth);
                     return;
                 }
-                ENFORCE(a.targs.size() == 1);
-                result = recursivelyFlattenArrays(gs, a.targs.front(), newDepth);
+
+                // If we're dealing with some other generic and it supports #to_ary then we can bypass it and just go
+                // straight to the return type of that method.
+                DispatchArgs innerArgs{Names::toAry(), args.locs, 0,
+                                       args.args,      type,      {type, args.fullType.origins},
+                                       type,           nullptr,   args.originForUninitialized};
```

Comment:
Likewise here; the code as written dispatches `to_ary` with the original arguments of the `flatten` call.